src/Simulation/Native/src/simulator/capi.cpp

```diff
@@ -216,6 +221,18 @@ extern "C"
         return Microsoft::Quantum::Simulator::get(id)->dumpQubits(qs, callback);
     }
 
+    MICROSOFT_QUANTUM_DECL bool DumpQubitsToLocation(
+        _In_ unsigned id,
+        _In_ unsigned n,
+        _In_reads_(n) unsigned* q,
+        _In_ TDumpToLocationCallback callback, 
+        _In_ TDumpLocation location)
```

Comment:
I'm a little confused about the API for this export; in particular, why does this function need both the location and a callback that takes that location? Given that the callbacks provided for this argument are all lambdas, I'd expect the  location to be captured in a closure cell of that lambda, rather than explicitly built into the API here. If you went with that approach, I don't think you would need this and the other new API declaration to be added to the native simulator, as you would then again have a callback of type `bool (*)(size_t, double, double)`.

Reply:
I agree with your idea, and that was the first thing I considered (and as I now understand, I probably should have tried that, at least out of curiosity). However,

There are a few topics in this decision.
1. We are crossing the dynamic lib boundary (ABI boundary) during these calls. Passing the C++ instances over this boundary is [unreliable](https://stackoverflow.com/questions/22797418/how-do-i-safely-pass-objects-especially-stl-objects-to-and-from-a-dll/22797419#22797419). That is why the callback is passed as a _function pointer_ (typical/reliable in C ABI), rather than [`std::function()`](http://www.cplusplus.com/reference/functional/function/?kw=function).
2. The lambda that _has no captures_, is (and/or can be implicitly cast to) the function pointer. That's what happens here. But if the lambda _does have captures_, then it cannot be represented/passed just as a function pointer. It becomes a dual entity - the function pointer + the reference to the captured data, i.e. it becomes a _function object_, sometimes called _functor_ (speaking in C++). Such an object cannot be just passed as an argument corresponding to the parameter of function pointer type.

That's what I'm doing here - passing the function object - "lambda with captures" - as 2 separate pointers explicitly.

---
I looked at how C# handles this. And for now I failed to figure out how they _magically_ pass a lambda with captures to this function pointer in
```c++
MICROSOFT_QUANTUM_DECL void Dump(_In_ unsigned id, _In_ bool (*callback)(size_t, double, double))
```
I realize that my understanding described above can be mistaken. Any explanation is welcome. Without the reliable explanation, I'm afraid passing a lambda with captures as a function pointer can be a hack that can break upon the next compiler upgrade. And [this](https://stackoverflow.com/questions/7852101/c-lambda-with-captures-as-a-function-pointer) makes me think that it will hardly compile at all.

_Requesting for an explanation/example how to implement better_.